cf/data/dask_regrid.py

```diff
@@ -0,0 +1,590 @@
+        # A reference source grid mask has already been incorporated
+        # into the weights matrix. Therefore, the mask for all slices
+        # of 'a' must be the same as this reference mask.
+        if variable_mask or (src_mask is None and ref_src_mask.any()):
+            raise ValueError(
+                f"Can't regrid with the {method!r} method when the source "
+                f"data mask varies over different {len(src_shape)}-d "
+                "regridding slices"
+            )
+
+        if ref_src_mask.dtype != bool or (
+            ref_src_mask.shape and ref_src_mask.shape != src_shape
+        ):
+            raise ValueError(
+                f"For {method!r} regridding, "
+                "the 'ref_src_mask' parameter must be None or a "
+                f"Boolean numpy array with shape () or {src_shape}. Got: "
+                f"dtype={ref_src_mask.dtype}, shape={ref_src_mask.shape}"
+            )
+
+        if src_mask is not None:
+            ref_src_mask = ref_src_mask.reshape(src_mask.shape)
+            if (src_mask != ref_src_mask).any():
+                raise ValueError(
+                    f"Can't regrid with the {method!r} method when the "
+                    "source data mask does not match the mask used to "
+                    "construct the regrid operator"
+                )
+
+    # ----------------------------------------------------------------
+    # Regrid the source data
+    # ----------------------------------------------------------------
+    if min_weight is None:
+        min_weight = np.finfo("float64").eps * 2.5
+
+    if variable_mask:
+        # Source data is masked and the source mask varies across
+        # slices => we have to regrid each slice separately, adjusting
+        # the weights for the mask of each slice.
+        #
+        # However, if the mask of a regrid slice is the same as the
+        # mask of its previous regrid slice then we can reuse the
+        # weights that already have the correct mask.
+        n_slices = a.shape[1]
+        regridded_data = np.ma.empty(
+            (dst_size, n_slices), dtype=weights.dtype, order="F"
+        )
+        prev_weights, prev_mask = None, None
+        for n in range(n_slices):
+            n = slice(n, n + 1)
+            a_n = a[:, n]
+            regridded_data[:, n], prev_mask, prev_weights = _regrid(
+                a_n,
+                np.ma.getmaskarray(a_n[:, 0]),
+                weights,
+                method,
+                prev_mask=prev_mask,
+                prev_weights=prev_weights,
+                min_weight=min_weight,
+            )
+
+        a = regridded_data
+        del a_n, regridded_data, prev_weights, prev_mask
+    else:
+        # Source data is either not masked or the source mask is same
+        # for all slices => all slices can be regridded
+        # simultaneously.
+        a, _, _ = _regrid(a, src_mask, weights, method, min_weight=min_weight)
+        del _
```

Comment:
If we want the second element above to be cleaned up as well, I think it'll need to be assigned a new name (unless it being overwritten as a variable by the third element will garbage collect it automatically, I'm not sure off the top of my head?).

Reply:
I think that the second `_` sends the first to GC, and the the `del` sends the second one to GC. It's all probably overkill, but doesn't harm, I guess.